drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2262 @@
+      (*addr) += close_ate.offset;
+      return 0;
+    }
+
+  /* The close_ate was invalid, `lets find out the last valid ate
+   * and point the address to this found ate.
+   *
+   * remark: if there was absolutely no valid data in the block *addr
+   * is kept at block_end - 2*ate_size, the next read will contain
+   * invalid data and continue with a block jump
+   */
+
+  return nvs_recover_last_ate(fs, addr);
+}
+
+/****************************************************************************
+ * Name: nvs_block_advance
+ ****************************************************************************/
+
+static void nvs_block_advance(FAR struct nvs_fs *fs, FAR uint32_t *addr)
+{
+  *addr += (1 << ADDR_BLOCK_SHIFT);
+  if ((*addr >> ADDR_BLOCK_SHIFT) == fs->geo.neraseblocks)
+    {
+      *addr -= (fs->geo.neraseblocks << ADDR_BLOCK_SHIFT);
+    }
+}
+
+/****************************************************************************
+ * Name: nvs_block_close
+ *
+ * Description:
+ *   allocation entry close (this closes the current block) by writing
+ *   offset of last ate to the block end.
+ *
+ ****************************************************************************/
+
+static int nvs_block_close(FAR struct nvs_fs *fs)
+{
+  int rc;
+  struct nvs_ate close_ate;
+
+  memset(&close_ate, CONFIG_MTD_CONFIG_ERASEDVALUE, sizeof(close_ate));
+  close_ate.id = NVS_SPECIAL_ATE_ID;
+  close_ate.len = 0;
+  close_ate.key_len = 0;
+  close_ate.offset =
+    (fs->ate_wra + sizeof(struct nvs_ate)) & ADDR_OFFS_MASK;
+
+  fs->ate_wra &= ADDR_BLOCK_MASK;
+  fs->ate_wra += fs->geo.erasesize - sizeof(struct nvs_ate);
+
+  nvs_ate_crc8_update(&close_ate);
+
+  rc = nvs_flash_ate_wrt(fs, &close_ate);
+  if (rc < 0)
+    {
+      ferr("Write ate failed, rc=%d\n", rc);
+    }
+
+  nvs_block_advance(fs, &fs->ate_wra);
+
+  fs->data_wra = fs->ate_wra & ADDR_BLOCK_MASK;
+  finfo("block close, data_wra=0x%" PRIx32 "\n", fs->data_wra);
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: nvs_add_gc_done_ate
+ ****************************************************************************/
+
+static int nvs_add_gc_done_ate(FAR struct nvs_fs *fs)
+{
+  struct nvs_ate gc_done_ate;
+
+  finfo("Adding gc done ate at %" PRIx32 "\n", fs->ate_wra & ADDR_OFFS_MASK);
+  memset(&gc_done_ate, CONFIG_MTD_CONFIG_ERASEDVALUE, sizeof(gc_done_ate));
+  gc_done_ate.id = NVS_SPECIAL_ATE_ID;
+  gc_done_ate.len = 0;
+  gc_done_ate.key_len = 0;
+  gc_done_ate.offset = fs->data_wra & ADDR_OFFS_MASK;
+  nvs_ate_crc8_update(&gc_done_ate);
+
+  return nvs_flash_ate_wrt(fs, &gc_done_ate);
+}
+
+/****************************************************************************
+ * Name: nvs_expire_ate
+ ****************************************************************************/
+
+static int nvs_expire_ate(FAR struct nvs_fs *fs, uint32_t addr)
+{
+  uint8_t expired;
+
+  expired = (uint8_t)~CONFIG_MTD_CONFIG_ERASEDVALUE;
```

Comment:
Simple id will need hash check logic  in startup/list, because of hash conflicts.
 I don't want to introduce more computing cost.
I will make sure all necessary checks are there.